tests/bluetooth/bsim_bt/bsim_test_mesh/src/test_friendship.c

```diff
@@ -815,33 +836,51 @@ static void test_other_group(void)
 	PASS();
 }
 
-#define TEST_CASE(role, name, description)                                     \
-	{                                                                      \
-		.test_id = "friendship_" #role "_" #name,                      \
-		.test_descr = description,                                     \
-		.test_post_init_f = test_##role##_init,                        \
-		.test_tick_f = bt_mesh_test_timeout,                           \
-		.test_main_f = test_##role##_##name,                           \
+/** LPN disable test.
+ *
+ * Check that toggling lpn_set() results in correct disabled state
+ */
+static void test_lpn_disable(void)
+{
+	bt_mesh_test_setup();
+
+	bt_mesh_lpn_set(true);
+	bt_mesh_lpn_set(false);
```

Comment:
While I was looking at this test, I found that in the lpn code, the terminated callback can be called even if the established callback hasn't been called yet. This happens when disabling the lpn between friend offer and friend update messages. The terminated callback is called regardless of whether the friendship was established or not (https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L230-L247). This can be reproduced with your test if add some delay (for me that was 400msec) before disabling lpn and commenting out lines 378-380. For me this looks like a bug.

Reply:
So what you are saying is that the terminated CB(s) should not trigger unless the LPN node actually has a established connection with a friendship node?

Reply:
For me that sounds logically correct.

Reply:
Did you get the actual babSim terminated CB to trigger? To me it seems like line https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L244 prevents the CB from triggering unless there was a friend node connected?

Reply:
`lpn->frnd` gets assigned when Friend Offer is received: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L574. At this point friendship is not established yet (LPN needs to wait for Friend Update: https://github.com/zephyrproject-rtos/zephyr/blob/main/subsys/bluetooth/mesh/lpn.c#L1004).

This is the log when I changed the test as I mentioned in the comment above:

> d_00: @00:00:00.000000  *** Booting Zephyr OS build v2.7.99-3008-gefdb18419891  ***
d_01: @00:00:00.000000  *** Booting Zephyr OS build v2.7.99-3008-gefdb18419891  ***
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_00: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.7 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Platform: Nordic Semiconductor (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: HW Variant: nRF52x (0x0002)
d_01: @00:00:00.000000  [00:00:00.000,000] <inf> bt_hci_core: Firmware: Standard Bluetooth controller (0x00) Version 2.7 Build 99
d_01: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.000000  [00:00:00.000,000] <wrn> bt_id: No static addresses stored in controller
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:3B:20:15:18:12 (random)
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Bluetooth initialized
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: Identity: ED:71:8F:C2:E4:6E (random)
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: HCI: version 5.3 (0x0c) revision 0x0000, manufacturer 0x0059
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_hci_core: LMP: version 5.3 (0x0c) subver 0xffff
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Bluetooth initialized
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_00: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> mesh_test: Mesh initialized
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0004
d_01: @00:00:00.002648  [00:00:00.002,624] <dbg> bt_mesh_main.bt_mesh_provision: net_idx 0x0000 flags 0x00 iv_index 0x0000
d_01: @00:00:00.302276  [00:00:00.302,246] <inf> test_friendship: LPN: Polling 0x0001 (initial)
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: Poll from 0x0004
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: established with 0x0004
d_00: @00:00:00.339845  [00:00:00.339,843] <wrn> bt_mesh_friend: Another request before previous completed sending
d_01: @00:00:00.395508  [00:00:00.395,507] <inf> test_friendship: LPN: terminated with 0x0001
d_00: @00:00:29.300782 INFO: test_friend_no_est PASSED
d_01: @00:00:29.693360 INFO: test_lpn_disable PASSED
d_00: @00:00:29.802735  [00:00:29.802,734] <wrn> bt_mesh_friend: Friendship lost with 0x0004
d_00: @00:00:29.802735  [00:00:29.802,734] <inf> test_friendship: Friend: terminated with 0x0004

There is a message `LPN: terminated with 0x0001`, but there is no preceding message `LPN: established ...`.

I think this could be fixed in a separate PR.

Reply:
Ok, I had to tweak my sleep time but i got the same now. It is clearly an issue:

> d_00: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0001
d_01: @00:00:00.002648  [00:00:00.002,624] <inf> bt_mesh_main: Primary Element: 0x0004
d_01: @00:00:00.003888  [00:00:00.003,875] <dbg> bt_mesh_lpn.lpn_set_state: disabled -> enabled
d_01: @00:00:00.003889  [00:00:00.003,875] <dbg> bt_mesh_lpn.send_friend_req: 
d_01: @00:00:00.004088  [00:00:00.004,058] <dbg> bt_mesh_lpn.friend_req_sent: 
d_01: @00:00:00.004088  [00:00:00.004,058] <dbg> bt_mesh_lpn.lpn_set_state: enabled -> req wait
d_01: @00:00:00.102540  [00:00:00.102,539] <dbg> bt_mesh_lpn.lpn_timeout: state: req wait
d_01: @00:00:00.102540  [00:00:00.102,539] <dbg> bt_mesh_lpn.lpn_set_state: req wait -> wait offer
d_01: @00:00:00.302225  [00:00:00.302,215] <dbg> bt_mesh_lpn.bt_mesh_lpn_friend_offer: recv_win 255 queue_size 16 sub_list_size 3 rssi -35 counter 0
d_01: @00:00:00.302225  [00:00:00.302,215] <dbg> bt_mesh_lpn.send_friend_poll: lpn->sent_req 0x00
d_01: @00:00:00.302735  [00:00:00.302,734] <dbg> bt_mesh_lpn.clear_friendship: force 0 disable 1, current state: wait offer
d_01: @00:00:00.302735  [00:00:00.302,734] <dbg> bt_mesh_lpn.lpn_set_state: wait offer -> disabled
d_01: @00:00:00.302735  [00:00:00.302,734] <inf> test_friendship: LPN: terminated with 0x0001
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: Poll from 0x0004
d_00: @00:00:00.306702  [00:00:00.306,701] <inf> test_friendship: Friend: established with 0x0004
d_00: @00:00:00.339845  [00:00:00.339,843] <wrn> bt_mesh_friend: Another request before previous completed sending
d_00: @00:00:29.802735  [00:00:29.802,734] <wrn> bt_mesh_friend: Friendship lost with 0x0004
d_00: @00:00:29.802735  [00:00:29.802,734] <inf> test_friendship: Friend: terminated with 0x0004

The LPN is disabled while it is waiting for an offer, even though it already has received one from the friend node...

Yes, I agree that this should be a separate PR. Do you have a clear idea how to fix it, or should i give it a go?

Reply:
> Do you have a clear idea how to fix it, or should i give it a go?

Not at the moment. You could give it a go if you have a time, or it can be investigated later.

Reply:
That's a good catch, I agree that this shouldn't happen. Can be a separate PR though